### PR TITLE
Roundend report has a meter for showing relative escape numbers

### PR DESCRIPTION
**Pull request by etherware-novice**
## About The Pull Request
Roundend report has a ascii bar for showing the relative number of people to escape on shuttle, escape on pod, be stranded, or die in the round

image
## Why It's Good For The Game
Fancy graphic, lets you see at a glance how much of a bloodbath the round was or how proactive people were at getting to the shuttle.
## Changelog
:cl:
add: roundend survival meter
/:cl:

### CODE REVIEW

**Comment by optimumtact** [comment]
I don't believe you

File: code/__HELPERS/roundend.dm [hunk 1]
```diff
@@ -4,6 +4,9 @@
 #define PERSONAL_LAST_ROUND "personal last round"
 #define SERVER_LAST_ROUND "server last round"
 
+#define ROUNDEND_BAR_SIZE 20
```
Comment (Mothblocks) [inline code/__HELPERS/roundend.dm:7]: undef at the end

File: code/__HELPERS/roundend.dm [hunk 2]
```diff
@@ -330,11 +333,33 @@
 	parts += "[FOURSPACES]Station Integrity: <B>[GLOB.station_was_nuked ? span_redtext("Destroyed") : "[popcount["station_integrity"]]%"]</B>"
 	var/total_players = GLOB.joined_player_list.len
 	if(total_players)
+		/// the proportions used for the rendered list
+		var/barSections = list()
```
Comment (Mothblocks) [inline code/__HELPERS/roundend.dm:337]: snake_case

File: code/__HELPERS/roundend.dm [hunk 3]
```diff
@@ -330,11 +333,33 @@
...
 		parts+= "[FOURSPACES]Total Population: <B>[total_players]</B>"
 		if(station_evacuated)
 			parts += "<BR>[FOURSPACES]Evacuation Rate: <B>[popcount[POPCOUNT_ESCAPEES]] ([PERCENT(popcount[POPCOUNT_ESCAPEES]/total_players)]%)</B>"
 			parts += "[FOURSPACES](on emergency shuttle): <B>[popcount[POPCOUNT_SHUTTLE_ESCAPEES]] ([PERCENT(popcount[POPCOUNT_SHUTTLE_ESCAPEES]/total_players)]%)</B>"
+
+			barSections += list( \
+			popcount[POPCOUNT_ESCAPEES], \
+			popcount[POPCOUNT_ESCAPEES] - popcount[POPCOUNT_SHUTTLE_ESCAPEES])  // non shuttle escapees
```
Comment (Mothblocks) [inline code/__HELPERS/roundend.dm:345]:
```suggestion
			barSections += list(
				popcount[POPCOUNT_ESCAPEES],
				popcount[POPCOUNT_ESCAPEES] - popcount[POPCOUNT_SHUTTLE_ESCAPEES],  // non shuttle escapees
			)
```

File: code/__HELPERS/roundend.dm [hunk 4]
```diff
@@ -330,11 +333,33 @@
...
+
+		else
+			barSections += list(0, 0) // 0 escapees, those fields are empty
+
 		parts += "[FOURSPACES]Survival Rate: <B>[popcount[POPCOUNT_SURVIVORS]] ([PERCENT(popcount[POPCOUNT_SURVIVORS]/total_players)]%)</B>"
+		barSections += (popcount[POPCOUNT_SURVIVORS] - popcount[POPCOUNT_ESCAPEES])  // stranded
+		barSections += (total_players - popcount[POPCOUNT_SURVIVORS])  // dead
+
+		var/list/renderedBar = generate_aggregate_bar(ROUNDEND_BAR_SIZE, barSections, totalSize = total_players)
```
Comment (Mothblocks) [inline code/__HELPERS/roundend.dm:354]: snake_case

File: code/__HELPERS/roundend.dm [hunk 5]
```diff
@@ -330,11 +333,33 @@
...
+
+		parts += "<B>( [join_color_list(renderedBar, ROUNDEND_BAR_COLOR)] )</B>"
+		parts += "<B>\
+		<span class=[ROUNDEND_BAR_COLOR[1]]>Shuttle</span>:\
+		<span class=[ROUNDEND_BAR_COLOR[2]]>No-Shuttle Escape</span>:\
+		<span class=[ROUNDEND_BAR_COLOR[3]]>Stranded</span>:\
+		<span class=[ROUNDEND_BAR_COLOR[4]]>Dead</span></B>"
```
Comment (Mothblocks) [inline code/__HELPERS/roundend.dm:361]: indent one

File: code/__HELPERS/text.dm [hunk 6]
```diff
@@ -1022,6 +1022,48 @@ GLOBAL_LIST_INIT(binary, list("0","1"))
 		loadstring += i <= limit ? "█" : "░"
 	return "\[[loadstring]\]"
 
+
+/proc/generate_aggregate_bar(numSquares, list/sectionSize, totalSize=null)
```
Comment (Mothblocks) [inline code/__HELPERS/text.dm:1026]: autodoc. snake_case

File: code/__HELPERS/text.dm [hunk 7]
```diff
@@ -1022,6 +1022,48 @@ GLOBAL_LIST_INIT(binary, list("0","1"))
...
+	var/list/returnList = list()
```
Comment (Mothblocks) [inline code/__HELPERS/text.dm:1027]: snake_case

File: code/__HELPERS/text.dm [hunk 8]
```diff
@@ -1022,6 +1022,48 @@ GLOBAL_LIST_INIT(binary, list("0","1"))
...
+			continue
+		listSum += num
+
+	if(isnull(totalSize) || totalSize > listSum)
+		totalSize = listSum
+
+
+	for (var/section in sectionSize)
+		var/sectionRender = ""
+		var/repeatAmount = FLOOR(( section / totalSize ) * numSquares, 1)
```
Comment (Mothblocks) [inline code/__HELPERS/text.dm:1042]:
```suggestion
		var/repeatAmount = FLOOR((section / totalSize) * numSquares, 1)
```

File: code/__HELPERS/text.dm [hunk 9]
```diff
@@ -1022,6 +1022,48 @@ GLOBAL_LIST_INIT(binary, list("0","1"))
...
+		if(repeatAmount < 1)
+			returnList += ""
+			continue
+		for(var/i in 1 to repeatAmount)
+			sectionRender += "#"
+		returnList += sectionRender
```
Comment (Mothblocks) [inline code/__HELPERS/text.dm:1048]:
```suggestion
		return_list += repeat_string(repeat_amount, "#")
```

File: code/__HELPERS/text.dm [hunk 10]
```diff
@@ -1022,6 +1022,48 @@ GLOBAL_LIST_INIT(binary, list("0","1"))
...
+
+	if(listSum < totalSize)
+		var/sectionRender = ""
+		for(var/i in 1 to ((totalSize - listSum) / totalSize) * numSquares)  // yep this is an eyesore
+			sectionRender += "_"
+		returnList += sectionRender
```
Comment (Mothblocks) [inline code/__HELPERS/text.dm:1054]: repeat_string

File: code/__HELPERS/roundend.dm [hunk 11]
```diff
@@ -330,11 +333,33 @@
...
+			barSections += list( \
+			popcount[POPCOUNT_ESCAPEES], \
+			popcount[POPCOUNT_ESCAPEES] - popcount[POPCOUNT_SHUTTLE_ESCAPEES])  // non shuttle escapees
+
+		else
+			barSections += list(0, 0) // 0 escapees, those fields are empty
+
 		parts += "[FOURSPACES]Survival Rate: <B>[popcount[POPCOUNT_SURVIVORS]] ([PERCENT(popcount[POPCOUNT_SURVIVORS]/total_players)]%)</B>"
+		barSections += (popcount[POPCOUNT_SURVIVORS] - popcount[POPCOUNT_ESCAPEES])  // stranded
+		barSections += (total_players - popcount[POPCOUNT_SURVIVORS])  // dead
```
Comment (Mothblocks) [inline code/__HELPERS/roundend.dm:352]: no two spaces

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/__HELPERS/text.dm (hunk 9 above)
Comment (etherware-novice) [reply code/__HELPERS/text.dm:1048]: wow i wish i knew that existed before

File: code/__HELPERS/roundend.dm (hunk 11 above)
Comment (etherware-novice) [reply code/__HELPERS/roundend.dm:353]: for the comments?

Comment (Mothblocks) [reply code/__HELPERS/roundend.dm:353]: yes

**Comment by blackdav123** [comment]
ascii is SOUL

**Comment by Mothblocks** [comment]
i keep looking at this and it just looks ugly sorry

**Comment by etherware-novice** [comment]
you...do know..you can just ask me to redesign it right???

**Comment by etherware-novice** [comment]
@Mothblocks like what do you think would be a good design

**Comment by Mothblocks** [comment]
oranges did and you told him no

**Comment by etherware-novice** [comment]
??? i dont know how tgui works though, i couldve changed the character used to make it look more appealing

**Comment by Mothblocks** [comment]
but if its just an ascii thing its always going to look ugly thats why i closed it

**Comment by etherware-novice** [comment]
thats subjective also therws plenty of ascii bars in the game already

**Comment by Mothblocks** [comment]
There shouldn't be

Yes it's subjective and the job of the maintainer is to make subjective decisions

**Pull request closed without merging**